downloader/src/main/kotlin/vcs/Mercurial.kt

```diff
@@ -0,0 +1,154 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.downloader.vcs
+
+import ch.frankel.slf4k.*
+
+import com.here.ort.downloader.Main
+import com.here.ort.downloader.VersionControlSystem
+import com.here.ort.util.ProcessCapture
+import com.here.ort.util.log
+
+import java.io.File
+import java.io.IOException
+
+object Mercurial : VersionControlSystem() {
+
+    private const val extensionSparse = "sparse = "
+    private const val extensionLargeFiles = "largefiles = "
+
+    override fun getWorkingDirectory(vcsDirectory: File) =
+            object : WorkingDirectory(vcsDirectory) {
+                override fun isValid(): Boolean {
+                    val repositoryRoot = runMercurialCommand(workingDir, "root").stdout().trim()
+                    return workingDir.path.startsWith(repositoryRoot)
+                }
+
+                override fun getRemoteUrl() =
+                        runMercurialCommand(workingDir, "paths", "default").stdout().trimEnd()
+
+                override fun getRevision() = runMercurialCommand(workingDir, "id", "-i").stdout().trimEnd()
+
+                override fun getRootPath(path: File) = runMercurialCommand(workingDir, "root").stdout().trimEnd()
+
+                override fun getPathToRoot(path: File): String {
+                    val absolutePath = if (path.isAbsolute || path == workingDir) {
+                        path
+                    } else {
+                        workingDir.resolve(path)
+                    }
+
+                    return absolutePath.relativeTo(File(getRootPath(absolutePath))).path
+                }
+            }
+
+    override fun isApplicableProvider(vcsProvider: String) = vcsProvider.equals("mercurial", true)
+
+    override fun isApplicableUrl(vcsUrl: String) = ProcessCapture("hg", "identify", vcsUrl).exitValue() == 0
+
+    override fun download(vcsUrl: String, vcsRevision: String?, vcsPath: String?, version: String,
+                          targetDir: File): String {
+
+        val revisionCmdArgs = mutableListOf<String>()
+
+        // We cannot detect beforehand if the Large Files extension would be required, so enable it by default.
+        val extensionsList = mutableListOf(extensionLargeFiles)
+
+        if (vcsRevision != null) {
+            revisionCmdArgs.add("-r")
+            revisionCmdArgs.add(vcsRevision)
+        }
+
+        if (!vcsPath.isNullOrEmpty()) {
+            // Enable sparse extension.
+            extensionsList.add(extensionSparse)
+        }
+
+        runMercurialCommand(targetDir, "init")
+        File(targetDir, ".hg/hgrc")
+                .writeText("[paths]\n" +
```

Comment:
Just an idea: Maybe use a multi-line string here for easier reading? Ensure proper indentation (not too many leading spaces) in that case by actually looking at the generated file.